api.md: document device invite apis (#12064)
tailscale/tailscale: api.md

The response is 2xx on success. The response body is currently an empty JSON object.

## List device invites

> ```

Reviewer: ```suggestion
```http
```

GET /api/v2/device/{deviceID}/device-invites
```

List all share invites for a device.

### Parameters

#### `deviceID` (required in URL path)

The ID of the device.

### Request example

```
curl -X GET "https://api.tailscale.com/api/v2/device/11055/device-invites" \
-u "tskey-api-xxxxx:"
```

### Response

```json
[
  {
    "id": "12345",
    "created": "2024-05-08T20:19:51.777861756Z",
    "tailnetId": 59954,
    "deviceId": 11055,
    "sharerId": 22011,
    "allowExitNode": true,
    "email": "user@example.com",
    "lastEmailSentAt": "2024-05-08T20:19:51.777861756Z",
    "inviteUrl": "https://login.tailscale.com/admin/invite/<code>",
    "accepted": false
  },
  {
    "id": "12346",
    "created": "2024-04-03T21:38:49.333829261Z",
    "tailnetId": 59954,
    "deviceId": 11055,
    "sharerId": 22012,
    "inviteUrl": "https://login.tailscale.com/admin/invite/<code>",
    "accepted": true,
    "acceptedBy": {
      "id": 33223,
      "loginName": "someone@example.com",
      "profilePicUrl": ""
    }
  }
]
```

## Create device invites

> ```

Reviewer 2: Can we add the `http` identifier here? we have it on some of the others

> ```

Reviewer: ```suggestion
```http
```

POST /api/v2/device/{deviceID}/device-invites
```

Create new share invites for a device.

### Parameters

#### `deviceID` (required in URL path)

The ID of the device.

#### List of invite requests (required in `POST` body)

> 

Author: ```suggestion

Each invite request is an object with the following optional fields:

```

> - **`multiUse`:** Optionally specifies whether the invite should be able to be accepted more than once. When set to true, results in an invite that can be accepted up to 1000 times.
> - **`allowExitNode`:** Optionally specifies whether the invited user will be able to use the device as an exit node when the device is advertising as one.
> - **`email`:** Optionally specifies the email to which to send the created invite. If not set, an invite URL is generated and returned, but not sent out.

Reviewer: ```suggestion
- **`multiUse`:** (Optional) Specify whether the invite can be accepted more than once. When set to `true`, it results in an invite that can be accepted up to 1,000 times.
- **`allowExitNode`:** (Optional) Specify whether the invited user can use the device as an exit node when it advertises as one.
- **`email`:** (Optional) Specify the email to send the created invite. If not set, the endpoint generates and returns an invite URL (but doesn't send it out).
```

### Request example

> ```

Reviewer: ```suggestion
```sh
```

curl -X POST "https://api.tailscale.com/api/v2/device/11055/device-invites" \
-u "tskey-api-xxxxx:" \
-H "Content-Type: application/json" \
--data-binary '[{"multiUse": true, "allowExitNode": true, "email":"user@example.com"}]'
```

### Response

> ```json

Reviewer 2: Same here

Reviewer 2: Whoops, I messed up my comments. Can we use `jsonc` here? :D

> ```json

Reviewer: ```suggestion
```jsonc
```

```jsonc
{}
```

# Device invites

> A device invite is an invitation that shares a device with a user external to the device's tailnet

Author: ```suggestion
A device invite is an invitation that shares a device with an external user (a user not in the device's tailnet).
```

Each device invite has a unique ID that is used to identify the invite in API calls.

> All device invite IDs for a paricular device can be found by [listing all device invites for a device](#list-device-invites)

Author: ```suggestion
All device invite IDs for a particular device can be found by [listing device invites](#list-device-invites).
```

> Each device invite has a unique ID that is used to identify the invite in API calls.
> All device invite IDs for a paricular device can be found by [listing all device invites for a device](#list-device-invites)

Reviewer: ```suggestion
Each device invite has a unique ID that is used to identify the invite in API calls.
You can find all device invite IDs for a particular device by [listing all device invites for a device](#list-device-invites).
```

```

Retrieve the specified device invite.

### Parameters

#### `deviceInviteId` (required in URL path)

The ID of the device share invite.

### Request example

```sh
curl "https://api.tailscale.com/api/v2/device-invites/12346" \
  -u "tskey-api-xxxxx:"
```

### Response

```jsonc
{
  "id": "12346",
  "created": "2024-04-03T21:38:49.333829261Z",
  "tailnetId": 59954,
  "deviceId": 11055,
  "sharerId": 22012,
  "multiUse": true,
  "allowExitNode": true,
  "email": "user@example.com",
  "lastEmailSentAt": "2024-04-03T21:38:49.333829261Z",
  "inviteUrl": "https://login.tailscale.com/admin/invite/<code>",
  "accepted": false
}
```

## Delete device invite

```http
DELETE /api/v2/device-invites/{deviceInviteId}
```

Delete the specified device invite.

### Parameters

#### `deviceInviteId` (required in URL path)

The ID of the device share invite.

### Request example

```sh
curl -X DELETE "https://api.tailscale.com/api/v2/device-invites/12346" \
  -u "tskey-api-xxxxx:"
```

### Response

The response is 2xx on success. The response body is an empty JSON object.

## Resend device invite

```http
POST /api/v2/device-invites/{deviceInviteId}/resend
```

> Resend the specified device invite by email. This can only be used if the specified invite was originally created with an email specified.
> See [creating device invites for a device](#create-device-invites).
> 
> Note that invite resends are rate limited to once per minute.

Reviewer: ```suggestion
Resend the specified device invite by email. You can only use this if the specified invite was originally created with an email specified. Refer to [creating device invites for a device](#create-device-invites).

Note: Invite resends are rate limited to one per minute.
```

### Parameters

#### `deviceInviteId` (required in URL path)

The ID of the device share invite.

### Request example

```sh
curl -X POST "https://api.tailscale.com/api/v2/device-invites/12346/resend" \
  -u "tskey-api-xxxxx:"
```

### Response

The response is 2xx on success. The response body is an empty JSON object.

## Accept device invite

```http
POST /api/v2/device-invites/-/accept
```

> Resend the specified device invite by email. This can only be used if the specified invite was originally created with an email specified.
> See [creating device invites for a device](#create-device-invites).
> 
> Note that invite resends are rate limited to once per minute.

Reviewer: ```suggestion
Resend the specified device invite by email. You can only use this if the specified invite was originally created with an email specified. Refer to [creating device invites for a device](#create-device-invites).

Note: Invite resends are rate limited to one per minute.
```

### Parameters

#### `invite` (required in `POST` body)

> Accepted as the URL of of the invite the form "https://login.tailscale.com/admin/invite/{code}"
> or just the "{code}" component of the above URL.

Reviewer: ```suggestion
The URL of the invite (in the form "https://login.tailscale.com/admin/invite/{code}") or the "{code}" component of the above URL.
```
